### ISSUE TITLE
False positive `clippy::redundant_closure_call` when using async closures as async try block

### ISSUE THREAD

**Issue by StackOverflowExcept1on** [rust-lang/rust-clippy#16232]
### Summary

Here's an example where you might want this (though this is a synchronous context):
https://github.com/dtolnay/sha1dir/blob/8c8946dec556bd26fab88d67662a7841b573bef4/src/main.rs#L146

### Lint Name

`clippy::redundant_closure_call`

### Reproducer

I tried this code:

```rust
#[tokio::main]
async fn main() {
    (async || -> Result<(), &'static str> { Ok(()) })()
        .await
        .unwrap();
}
```

I saw this happen:

```
    Checking playground v0.0.1 (/playground)
warning: try not to call a closure in the expression where it is declared
 --> src/main.rs:3:5
  |
3 |     (async || -> Result<(), &'static str> { Ok(()) })()
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try doing something like: `async { Ok(()) }`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/rust-1.92.0/index.html#redundant_closure_call
  = note: `#[warn(clippy::redundant_closure_call)]` on by default

warning: `playground` (bin "playground") generated 1 warning (run `cargo clippy --fix --bin "playground" -p playground` to apply 1 suggestion)
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.65s
```

I expected to see this happen: this is valid code and it doesn't cause any problems since you just want to make an asynchronous try block.

### Version

```text
rustc 1.92.0-nightly (f04e3dfc8 2025-10-19)
binary: rustc
commit-hash: f04e3dfc87d7e2b6ad53e7a52253812cd62eba50
commit-date: 2025-10-19
host: x86_64-unknown-linux-gnu
release: 1.92.0-nightly
LLVM version: 21.1.3
```

**Comment by samueltardieu** [comment]
I am not sure I follow why this would be useful. However, I would understand not wanting to lint a closure of any kind when an early return is used inside the closure, which is the case in the link you give since it might return early because of `?`. Is that what you have in mind? (and it looks like this case is already properly handled)

**Comment by aleksanderkrauze** [comment]
Hi. I would like to give this issue some more visibility, as I have just triggered it myself.

To give more context: currently synchronous version of this example passes and `clippy::redundant_closure_call` is not raised:

```rust
use std::io::Error;

fn foo() -> Result<(), Error> {
    Ok(())
}

fn bar() -> Result<(), Error> {
    Ok(())
}

fn main() {
    let x: Result<(), Error> = (|| {
        let () = foo()?;
        let () = bar()?;
        
        Ok(())
    })();
    
    println!("{x:?}");
}
```

However asynchronous version using async closure, or closure returning async block triggers this lint:

```rust
use std::io::Error;

async fn foo() -> Result<(), Error> {
    Ok(())
}

async fn bar() -> Result<(), Error> {
    Ok(())
}

#[tokio::main]
async fn main() {
    let x: Result<(), Error> = (async || {
        let () = foo().await?;
        let () = bar().await?;
        
        Ok(())
    })().await;
    
    println!("{x:?}");
    
    let y: Result<(), Error> = (|| async {
        let () = foo().await?;
        let () = bar().await?;
        
        Ok(())
    })().await;
    
    println!("{y:?}");
}
```
<details>
<summary>clippy's output</summary>

```
    Checking playground v0.0.1 (/playground)
warning: try not to call a closure in the expression where it is declared
  --> src/main.rs:13:32
   |
13 |       let x: Result<(), Error> = (async || {
   |  ________________________________^
14 | |         let () = foo().await?;
15 | |         let () = bar().await?;
...  |
18 | |     })().await;
   | |________^
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/rust-1.94.0/index.html#redundant_closure_call
   = note: `#[warn(clippy::redundant_closure_call)]` on by default
help: try doing something like
   |
13 ~     let x: Result<(), Error> = async {
14 +         let () = foo().await?;
15 +         let () = bar().await?;
16 +         
17 +         Ok(())
18 ~     }.await;
   |

warning: try not to call a closure in the expression where it is declared
  --> src/main.rs:22:32
   |
22 |       let y: Result<(), Error> = (|| async {
   |  ________________________________^
23 | |         let () = foo().await?;
24 | |         let () = bar().await?;
...  |
27 | |     })().await;
   | |________^
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/rust-1.94.0/index.html#redundant_closure_call
help: try doing something like
   |
22 ~     let y: Result<(), Error> = async {
23 +         let () = foo().await?;
24 +         let () = bar().await?;
25 +         
26 +         Ok(())
27 ~     }.await;
   |
```
</details>

In my opinion this is a false positive. Invoking closure immediately is a know-pattern and workaround for lack of `try` blocks. Clippy correctly detects it and does not raise `clippy::redundant_closure_call` in the case of synchronous code. There is no reason to treat async closures (or closures returning async blocks) any different.

**Comment by lkirkwood** [comment]
This definitely seems incorrect given the following code only lints the async version.
```rust
#![allow(unused)]

fn maybe() -> Option<()> {
    Some(())
}

async fn test() {
    let result = (|| {
        if maybe().is_none() {
            return Some(0);
        }

        Some(42)
    })();

    let async_result = (async || {
        if maybe().is_none() {
            return Some(0);
        }

        Some(42)
    })()
    .await;
}

fn main() {}
```

```
cargo clippy --all-targets --all-features 
    Checking playground v0.1.0 (/home/linus/tmp/playground)
warning: try not to call a closure in the expression where it is declared
  --> src/main.rs:16:24
   |
16 |       let async_result = (async || {
   |  ________________________^
17 | |         if maybe().is_none() {
18 | |             return Some(0);
...  |
21 | |         Some(42)
22 | |     })()
   | |________^
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/rust-1.94.0/index.html#redundant_closure_call
   = note: `#[warn(clippy::redundant_closure_call)]` on by default
help: try doing something like
   |
16 ~     let async_result = async {
17 +         if maybe().is_none() {
18 +             return Some(0);
19 +         }
20 + 
21 +         Some(42)
22 +     }
   |
```

@rustbot claim
